### PR TITLE
Do `gramine-sgx-get-token` automatically

### PR DESCRIPTION
**Pull request by woju**
## Description of the changes 

There's no EINITTOKEN on upstream driver (nor DCAP), and most people use that driver (it is/will be the only packaged flavour). It's bad UX to require them to do `gramine-sgx-get-token` for nothing (like, 384 * `\0`).

This PR is the first of series of two. It changes UX (`gramine-sgx-get-token` gets called automatically), but doesn't change behaviour of loader (the `.token` file is still required).

## How to test this PR? 

Run `gramine-sgx` without doing `gramine-sgx-get-token` first.

- - -
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)

### CODE REVIEW

**Review by oshogbo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)* status: 0 of 17 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1093#-NKx3mIn4WJOcvfRK0ZK:-NKx3mIn4WJOcvfRK0ZL:b-85i63v) (no related file):*
LGTM.

___
___
*[`Documentation/manpages/gramine-sgx-get-token.rst` line 15 at r1](https://reviewable.io/reviews/gramineproject/gramine/1093#-NKx3aBG2JdvWoTT2gCH:-NKx3aBH3Vne5rIvrNpg:bk1xdmt) ([raw file](https://github.com/gramineproject/gramine/blob/1afb25dac3063b9b3e791d03632e1350b3b83cfb/Documentation/manpages/gramine-sgx-get-token.rst#L15)):*
> ```reStructuredText
> 
> Using this command is not necessary (it was previously), since the token is
> fetched automatically if needed during first enclave start.
> ```

If this command is not to be used by user - maybe we can just remove this man page?
Or maybe we should add some note to remove it after 2 version.

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 17 of 17 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1BM1H46P-SW8aA2T7:-NL1BM1H46P-SW8aA2T8:b2q9pvp) (no related file):*
Same should be done for:
- https://github.com/gramineproject/examples
- https://github.com/gramineproject/gsc

But only after this PR is merged and appears in the release of Gramine.

@woju Do you want to do these PRs, or should we just create corresponding GitHub issues, or should somebody else (=me?) create such PRs?

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1CldL9Q1m8II_XWXs:-NL1CldL9Q1m8II_XWXt:b-xak97i) (no related file):*
Jenkins failed.

___
___
*[`Documentation/manpages/gramine.rst` line 30 at r1](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1C2sT-VTX0ocfv2gv:-NL1C2sT-VTX0ocfv2gw:b-w9be) ([raw file](https://github.com/gramineproject/gramine/blob/1afb25dac3063b9b3e791d03632e1350b3b83cfb/Documentation/manpages/gramine.rst#L30)):*
> ```reStructuredText
>    If not empty, :command:`gramine-sgx` will not automatically generate
>    EINITTOKEN (for out-of-tree driver), or its dummy counterpart (for upstream
>    and DCAP drivers).
> ```

Why do we introduce such a variable? I guess it doesn't hurt so I'm not blocking. But also, I don't see much need for it.
___
*[`Documentation/manpages/gramine-sgx-get-token.rst` line 15 at r1](https://reviewable.io/reviews/gramineproject/gramine/1093#-NKx3aBG2JdvWoTT2gCH:-NL1CIok4XvB1XXZPWdU:b2jstov) ([raw file](https://github.com/gramineproject/gramine/blob/1afb25dac3063b9b3e791d03632e1350b3b83cfb/Documentation/manpages/gramine-sgx-get-token.rst#L15)):*
<details><summary><i>Previously, oshogbo (Mariusz Zaborski) wrote…</i></summary><blockquote>

If this command is not to be used by user - maybe we can just remove this man page?
Or maybe we should add some note to remove it after 2 version.
</blockquote></details>

I wouldn't remove this man page and I wouldn't remove this command. It may be useful for users who still have non-FLC SGX machines (= they need the token), and these users may want to use this script explicitly in their flows, for whatever reason.

**Review by woju — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)* status: 16 of 17 files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv and @oshogbo)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1093#-NKx3mIn4WJOcvfRK0ZK:-NLLAM3N7vtsdjKqGel3:b-73n9g) (no related file):*
<details><summary><i>Previously, oshogbo (Mariusz Zaborski) wrote…</i></summary><blockquote>

LGTM.
</blockquote></details>

so plz unblock :)

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1CldL9Q1m8II_XWXs:-NLL93VT1vi9Y4OFH0LE:b6e5xgq) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Jenkins failed.
</blockquote></details>

Yeah, for stdout. Should be fixed.

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1BM1H46P-SW8aA2T7:-NLL9_I689BvM5RWaPtQ:b2j5ab2) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Same should be done for:
- https://github.com/gramineproject/examples
- https://github.com/gramineproject/gsc

But only after this PR is merged and appears in the release of Gramine.

@woju Do you want to do these PRs, or should we just create corresponding GitHub issues, or should somebody else (=me?) create such PRs?
</blockquote></details>

I'd prefer to make separate issues and fix when convenient. Those changes should be compatible, so if someone (gsc) is still calling `gramine-sgx-get-token`, it shouldn't break anything, so it's not that urgent to fix.

___
___
*[`Documentation/manpages/gramine.rst` line 30 at r1](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1C2sT-VTX0ocfv2gv:-NLL3BuL-NLKiEOmJDXL:b5nydrz) ([raw file](https://github.com/gramineproject/gramine/blob/1afb25dac3063b9b3e791d03632e1350b3b83cfb/Documentation/manpages/gramine.rst#L30)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Why do we introduce such a variable? I guess it doesn't hurt so I'm not blocking. But also, I don't see much need for it.
</blockquote></details>

As a rule, you should be able to undo any automation. For instance, automatic `get-token` adds latency (on the order of 2000 ms) to starting enclave (or alternatively it adds variance to enclave start time, depending on how you view it). So if someone would be actually measuring start times, I'd expect s/he'll `get-token` separately and set this variable as a safety device to make sure s/he's measuring what s/he meant, or fail hard.
___
*[`Documentation/manpages/gramine-sgx-get-token.rst` line 15 at r1](https://reviewable.io/reviews/gramineproject/gramine/1093#-NKx3aBG2JdvWoTT2gCH:-NLL8UTB7bAKPwiGqDTT:bib6gft) ([raw file](https://github.com/gramineproject/gramine/blob/1afb25dac3063b9b3e791d03632e1350b3b83cfb/Documentation/manpages/gramine-sgx-get-token.rst#L15)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I wouldn't remove this man page and I wouldn't remove this command. It may be useful for users who still have non-FLC SGX machines (= they need the token), and these users may want to use this script explicitly in their flows, for whatever reason.
</blockquote></details>

As long as the command is available, the manpage should remain.

**Comment by woju** [comment]
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Done:
- https://github.com/gramineproject/examples/issues/46
- https://github.com/gramineproject/gsc/issues/119
</blockquote></details>

Thanks!

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 16 of 17 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)* status: all files reviewed, 2 unresolved discussions (waiting on @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1BM1H46P-SW8aA2T7:-NLRs5bD-BTviRctmuxe:bat6x62) (no related file):*
> But only after this PR is merged and appears in the release of Gramine.

Isn't GSC master using Gramine master? We have GSC tagging for some reason :)

___
___
*[`Documentation/manpages/gramine.rst` line 30 at r1](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1C2sT-VTX0ocfv2gv:-NLRsZydBwWuBxtr66sF:b32hkgv) ([raw file](https://github.com/gramineproject/gramine/blob/1afb25dac3063b9b3e791d03632e1350b3b83cfb/Documentation/manpages/gramine.rst#L30)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

As a rule, you should be able to undo any automation. For instance, automatic `get-token` adds latency (on the order of 2000 ms) to starting enclave (or alternatively it adds variance to enclave start time, depending on how you view it). So if someone would be actually measuring start times, I'd expect s/he'll `get-token` separately and set this variable as a safety device to make sure s/he's measuring what s/he meant, or fail hard.
</blockquote></details>

Jeeez, why get-token takes 2 seconds? It's a local operation which doesn't require any networking, this sounds completely unreasonable.
// but this is unrelated to the PR, at least not directly
___
*[`Documentation/manpages/gramine.rst` line 29 at r2](https://reviewable.io/reviews/gramineproject/gramine/1093#-NLRoaGCDVESRkBT6jmQ:-NLRoaGCDVESRkBT6jmR:b-xi5fbx) ([raw file](https://github.com/gramineproject/gramine/blob/8738eb9521b193a78a253807a7cf320f5a907fe8/Documentation/manpages/gramine.rst#L29)):*
> ```reStructuredText
> 
>    If not empty, :command:`gramine-sgx` will not automatically generate
>    EINITTOKEN (for out-of-tree driver), or its dummy counterpart (for upstream
> ```

In official docs we need to be more precise, because AFAIR some people call non-upstreamed DCAP driver "out-of-tree driver" and the upstreamed one "in-tree driver".
@dimakuv: What's the best name to use here?



_Suggestion:_
```reStructuredText
for out-of-tree EPID driver
```
___
*[`Documentation/manpages/gramine-sgx-get-token.rst` line 15 at r2](https://reviewable.io/reviews/gramineproject/gramine/1093#-NLRqOSVE0Q5GxoSUrzP:-NLRqOSVE0Q5GxoSUrzQ:b-hdx5b) ([raw file](https://github.com/gramineproject/gramine/blob/8738eb9521b193a78a253807a7cf320f5a907fe8/Documentation/manpages/gramine-sgx-get-token.rst#L15)):*
> ```reStructuredText
> 
> Using this command is not necessary (it was previously), since the token is
> fetched automatically if needed during first enclave start.
> ```

(I think)



_Suggestion:_
```reStructuredText
during the first enclave start.
```

**Review by dimakuv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)* status: all files reviewed, 2 unresolved discussions (waiting on @mkow and @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1BM1H46P-SW8aA2T7:-NLV3xIDENAv9YGhRXO3:b6mb2ec) (no related file):*
> > But only after this PR is merged and appears in the release of Gramine.
> 
> Isn't GSC master using Gramine master? We have GSC tagging for some reason :)

I see no reason to break GSC like this. If we can avoid it (by postponing by a couple weeks), I prefer avoiding it.

But yes, technically no specific reason.

___
___
*[`Documentation/manpages/gramine.rst` line 30 at r1](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1C2sT-VTX0ocfv2gv:-NLV4mjj2jan3UROfbQ-:b-fdupw8) ([raw file](https://github.com/gramineproject/gramine/blob/1afb25dac3063b9b3e791d03632e1350b3b83cfb/Documentation/manpages/gramine.rst#L30)):*
> Jeeez, why get-token takes 2 seconds? It's a local operation which doesn't require any networking, this sounds completely unreasonable.

Firstly, this happens only for non-FLC systems. Secondly, iirc, producing a token (`EINITTOKEN`) requires *launching* the special Launch Enclave (LE) -- so basically we need to start an additional SGX enclave to get the token. And of course starting and then destroying an SGX enclave is slow.
___
*[`Documentation/manpages/gramine.rst` line 29 at r2](https://reviewable.io/reviews/gramineproject/gramine/1093#-NLRoaGCDVESRkBT6jmQ:-NLV4Bfn9pWLiWlqeQJA:bnx2sv1) ([raw file](https://github.com/gramineproject/gramine/blob/8738eb9521b193a78a253807a7cf320f5a907fe8/Documentation/manpages/gramine.rst#L29)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

In official docs we need to be more precise, because AFAIR some people call non-upstreamed DCAP driver "out-of-tree driver" and the upstreamed one "in-tree driver".
@dimakuv: What's the best name to use here?
</blockquote></details>

@mkow Your proposal sounds good. Adding `EPID` makes it unambigious. We also call it `old` and `deprecated` in our docs: https://gramine.readthedocs.io/en/latest/sgx-intro.html#linux-kernel-drivers. I also refer to it as `legacy`.

Anyway, Michal's proposal is good enough.

**Review by woju — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)* status: 15 of 17 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv and @mkow)

___
*[`Documentation/manpages/gramine.rst` line 30 at r1](https://reviewable.io/reviews/gramineproject/gramine/1093#-NL1C2sT-VTX0ocfv2gv:-NLW0tLO8H4l8I0YE8L2:b-16qv34) ([raw file](https://github.com/gramineproject/gramine/blob/1afb25dac3063b9b3e791d03632e1350b3b83cfb/Documentation/manpages/gramine.rst#L30)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

> Jeeez, why get-token takes 2 seconds? It's a local operation which doesn't require any networking, this sounds completely unreasonable.

Firstly, this happens only for non-FLC systems. Secondly, iirc, producing a token (`EINITTOKEN`) requires *launching* the special Launch Enclave (LE) -- so basically we need to start an additional SGX enclave to get the token. And of course starting and then destroying an SGX enclave is slow.
</blockquote></details>

Up to half of that (500-1000 ms) is Python interpreter start, which does slow things (like reading a number of small files). I think there's draft PR that RIIR's `gramine-sgx-get-token` exactly for this reason. Recent Python releases brought some improvements in this area, but we're talking about old driver on old kernel, which means old distro with old Python interpreters, so those improvements aren't available if we have `#!/usr/bin/python3` shebang.
___
*[`Documentation/manpages/gramine.rst` line 29 at r2](https://reviewable.io/reviews/gramineproject/gramine/1093#-NLRoaGCDVESRkBT6jmQ:-NLW4FXx6ewX2sJCxm0L:buhtp0c) ([raw file](https://github.com/gramineproject/gramine/blob/8738eb9521b193a78a253807a7cf320f5a907fe8/Documentation/manpages/gramine.rst#L29)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@mkow Your proposal sounds good. Adding `EPID` makes it unambigious. We also call it `old` and `deprecated` in our docs: https://gramine.readthedocs.io/en/latest/sgx-intro.html#linux-kernel-drivers. I also refer to it as `legacy`.

Anyway, Michal's proposal is good enough.
</blockquote></details>

Done. And I also need to fix this in #1094, because the note about "dummy counterpart" is not right.
___
*[`Documentation/manpages/gramine-sgx-get-token.rst` line 15 at r2](https://reviewable.io/reviews/gramineproject/gramine/1093#-NLRqOSVE0Q5GxoSUrzP:-NLW6mAdAhjexSlDjyHP:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/8738eb9521b193a78a253807a7cf320f5a907fe8/Documentation/manpages/gramine-sgx-get-token.rst#L15)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

(I think)
</blockquote></details>

Done.

**Review by dimakuv — DISMISSED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)* status: all files reviewed, 3 unresolved discussions (waiting on @mkow and @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1093#-NLWN8o53HBEpvoqhbWN:-NLWN8o53HBEpvoqhbWO:bhdlw9h) (no related file):*
`Jenkins-direct-sanitizers` failed.

___

**Review by woju — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)* status: all files reviewed, 3 unresolved discussions (waiting on @dimakuv and @mkow)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1093#-NLWN8o53HBEpvoqhbWN:-NLWOQHB9Oez1h0cajqB:be5osvx) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`Jenkins-direct-sanitizers` failed.
</blockquote></details>

Jenkins slave is heavily swapped, I'll fix it and rerun pipeline.

___

**Comment by woju** [comment]
Jenkins, test Jenkins-Direct-Sanitizers please

**Comment by woju** [comment]
Jenkins, test Jenkins-SGX-18.04-apps please

**Comment by woju** [comment]
Jenkins, test Jenkins-SGX-20.04-apps please

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)* status: all files reviewed, 1 unresolved discussion (waiting on @woju)

**Review by mkow — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1093)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**
